### PR TITLE
EkA 132: Creation of key material and decryption

### PR DESCRIPTION
**Pull request by Nexengineer**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -26,6 +26,7 @@ dependencies {
     implementation 'org.springframework:spring-context-support'
     implementation 'com.google.guava:guava:28.2-jre'
     compileOnly 'org.projectlombok:lombok'
+    compile group: 'org.bouncycastle', name: 'bcprov-jdk15on', version: '1.64'
```
Comment (ganesanarun) [inline build.gradle]: Follow same convention as other imports.

Comment (ganesanarun) [reply build.gradle]: and compile is deprecated in gradle 7, replace compile with Implementation.

File: src/main/java/in/org/projecteka/hiu/HiuConfiguration.java [hunk 2]
```diff
@@ -186,20 +187,29 @@ public HealthInformationRepository healthInformationRepository(PgPool pgPool) {
...
     public DataFlowRequestListener dataFlowRequestListener(MessageListenerContainerFactory messageListenerContainerFactory,
                                                            DestinationsConfig destinationsConfig,
                                                            DataFlowClient dataFlowClient,
-                                                           DataFlowRepository dataFlowRepository) {
+                                                           DataFlowRepository dataFlowRepository,
+                                                           CryptoHelper cryptoHelper) {
         return new DataFlowRequestListener(messageListenerContainerFactory,
                 destinationsConfig,
-                dataFlowClient, dataFlowRepository);
+                dataFlowClient, dataFlowRepository, cryptoHelper);
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]:
```suggestion
                dataFlowClient,
                dataFlowRepository,
                cryptoHelper);
```

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java [hunk 3]
```diff
@@ -49,6 +58,42 @@ public DataFlowRepository(PgPool pgPool) {
...
+                                DataFlowRequestKeyMaterial keyPairs = null;
+                                for (Row row : rows) {
+                                    JsonObject keyPairsJson = (JsonObject) row.getValue("key_pairs");
+                                    keyPairs = keyPairsJson.mapTo(DataFlowRequestKeyMaterial.class);
+                                }
+                                monoSink.success(keyPairs);
+                            }
+                        }
+                )
+        );
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java]:
```suggestion
                                    keyPairs = keyPairsJson.mapTo(DataFlowRequestKeyMaterial.class);
                                }
                                monoSink.success(keyPairs);
                            }}));
```

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java [hunk 4]
```diff
@@ -42,18 +47,55 @@ public void subscribe() throws ClientError {
...
+
+                logger.info("Initiating data flow request to consent manager");
+                dataFlowClient.initiateDataFlowRequest(dataFlowRequest)
+                        .flatMap(dataFlowRequestResponse ->
+                                dataFlowRepository.addDataRequest(dataFlowRequestResponse.getTransactionId(), dataFlowRequest)
+                                        .then(dataFlowRepository.addKeys(
+                                                dataFlowRequestResponse.getTransactionId(),
+                                                dataRequestKeyMaterial)))
+                        .block();
+            } catch (Exception exception){
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]:
```suggestion
            } catch (Exception exception) {
```

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java [hunk 5]
```diff
@@ -42,18 +47,55 @@ public void subscribe() throws ClientError {
...
+
+    private KeyMaterial keyMaterial(DataFlowRequestKeyMaterial dataFlowKeyMaterial) {
+        logger.info("Creating KeyMaterials");
+        return KeyMaterial.builder()
+                .cryptoAlg(CryptoHelper.ALGORITHM)
+                .curve(CryptoHelper.CURVE)
+                .dhPublicKey(KeyStructure.builder()
+                        .expiry("")
+                        .keyValue(dataFlowKeyMaterial.getPublicKey())
+                        .parameters("").build())
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]: Why expiry and parameters are empty?

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java [hunk 6]
```diff
@@ -42,18 +47,55 @@ public void subscribe() throws ClientError {
...
-            DataFlowRequest.builder().build();
-            logger.info("Initiating data flow request to consent manager");
-            dataFlowClient.initiateDataFlowRequest(dataFlowRequest)
-                    .flatMap(dataFlowRequestResponse ->
-                            dataFlowRepository.addDataRequest(dataFlowRequestResponse.getTransactionId(), dataFlowRequest))
-                    .block();
+            try{
+                var dataRequestKeyMaterial = dataFlowRequestKeyMaterial();
+                var keyMaterial = keyMaterial(dataRequestKeyMaterial);
+                dataFlowRequest.setKeyMaterial(keyMaterial);
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]: Instead of mutating the existing object, can you create a new object?

Comment (ganesanarun) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]: Hint: You could use toBuilder

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestRepository.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package in.org.projecteka.hiu.dataflow;
+
+import in.org.projecteka.hiu.dataflow.model.DataFlowRequest;
+import io.vertx.core.json.JsonObject;
+import io.vertx.pgclient.PgPool;
+import io.vertx.sqlclient.Tuple;
+import reactor.core.publisher.Mono;
+
+public class DataFlowRequestRepository {
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestRepository.java]: Don't we have one already?

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java [hunk 8]
```diff
@@ -2,23 +2,58 @@
...
+        return cryptoHelper.decrypt(keyPairs.getPrivateKey(),
+                senderPublicKey,
+                randomKeySender,
+                keyPairs.getRandomKey(),
+                entry.getContent());
+    }
 
     public Mono<Void> handleNotification(DataNotificationRequest dataNotificationRequest) {
         return Flux.fromIterable(dataNotificationRequest.getEntries())
                 .filter(entry -> entry.getLink() == null)
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: create a new method called isComponent which takes entry, and 

> return entry.getLink() == null
and use that in filter

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java [hunk 9]
```diff
@@ -2,23 +2,58 @@
...
+                .flatMap(entry -> getDecodedData(dataNotificationRequest, entry))
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: Do we need to make a DB call for each entry? Would not it better to get the keys from the DB once and use it for all the entries?

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java [hunk 10]
```diff
@@ -2,23 +2,58 @@
...
+                                .link(entry.getLink())
+                                .media(entry.getMedia())
+                                .build());
+                    } catch (Exception e) {
+                        logger.error("Error while decrypting {exception}", e);
+                        return Mono.error(e);
+                    }
+                });
+    }
+
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: Can you create a separate class/static method to handle the decryption which does not need to know where we are getting keys from, get all the values it needs to decrypt any entry given keys?

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/AesUtil.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+            int retLen = cipher.processBytes
+                    (encryptedBytes, 0, encryptedBytes.length, plainBytes, 0);
+            cipher.doFinal(plainBytes, retLen);
+
+            decryptedData = new String(plainBytes, Charset.forName("UTF-8"));
+        } catch (IllegalArgumentException | IllegalStateException |
+                DataLengthException | InvalidCipherTextException ex) {
+            System.out.println(ex.getMessage());
+        }
+        return decryptedData;
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/AesUtil.java]: this whole thing can be private method in your new class.

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        ECPublicKey eckey = (ECPublicKey)key;
+        return eckey.getQ().getEncoded(false);
+    }
+
+    public static byte [] savePrivateKey (PrivateKey key) throws Exception
+    {
+        ECPrivateKey eckey = (ECPrivateKey)key;
+        return eckey.getD().toByteArray();
+    }
+
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java]: Why it's called as saveX?

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        byte[] iv = Arrays.copyOfRange(xorOfRandoms, xorOfRandoms.length - 12, xorOfRandoms.length);
+
+        HKDFBytesGenerator hkdfBytesGenerator = new HKDFBytesGenerator(new SHA256Digest());
+        HKDFParameters hkdfParameters = new HKDFParameters(getBytesForBase64String(sharedKey), salt, null);
+        hkdfBytesGenerator.init(hkdfParameters);
+        byte[] aesKey = new byte[32];
+        hkdfBytesGenerator.generateBytes(aesKey, 0, 32);
+
+        return AesUtil.decrypt(getBytesForBase64String(encryptedMessage), aesKey, iv);
+    }
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java]: This should not be here.

File: src/main/resources/application-dev.yml [hunk 14]
```diff
@@ -1,19 +1,20 @@
...
+    url: http://localhost:8080/
   database:
     host: localhost
     port: 5432
     schema: health_information_user
     user: postgres
-    password: postgres
+    password: welcome
     poolSize: 5
 spring:
```
Comment (ganesanarun) [inline src/main/resources/application-dev.yml]: Does it need to be pushed?

File: build.gradle (hunk 1 above)
Comment (Nexengineer) [reply build.gradle]: Done

File: src/main/java/in/org/projecteka/hiu/HiuConfiguration.java (hunk 2 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java (hunk 3 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java (hunk 4 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java (hunk 6 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestListener.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestRepository.java (hunk 7 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRequestRepository.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java (hunk 8 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java (hunk 9 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: Nice catch! Fixed

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java (hunk 10 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/DataFlowService.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/AesUtil.java (hunk 11 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/AesUtil.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java (hunk 12 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java]: Done

File: src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java (hunk 13 above)
Comment (Nexengineer) [reply src/main/java/in/org/projecteka/hiu/dataflow/cryptohelper/CryptoHelper.java]: Removed

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java [hunk 15]
```diff
@@ -49,6 +58,39 @@ public DataFlowRepository(PgPool pgPool) {
...
+                            if (handler.failed())
+                                monoSink.error(dbOperationFailure("Failed to fetch consent request"));
+                            else {
+                                RowSet<Row> rows = handler.result();
+                                DataFlowRequestKeyMaterial keyPairs = null;
+                                for (Row row : rows) {
+                                    JsonObject keyPairsJson = (JsonObject) row.getValue("key_pairs");
+                                    keyPairs = keyPairsJson.mapTo(DataFlowRequestKeyMaterial.class);
+                                }
+                                monoSink.success(keyPairs);
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java:90]: Have to handle/check what would happen on this scenario.

File: src/main/java/in/org/projecteka/hiu/dataflow/Decryptor.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+        byte[] out = new byte[randomSender.length];
+        for (int i = 0; i < randomSender.length; i++) {
+            out[i] = (byte) (randomSender[i] ^ randomReceiver[i%randomReceiver.length]);
+        }
+        return out;
+    }
+
+    public String getBase64String(byte[] value){
+
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/Decryptor.java:122]: Fix this in next PR.

**Review by ganesanarun — APPROVED** [review]

**Pull request merged**